sample-app/src/components/InputDropdown.tsx

```diff
@@ -52,71 +48,114 @@ export default function InputDropdown({
+    focusedSectionIndex: undefined,
     shouldDisplayDropdown: false,
   });
-  const focusOptionId = focusedOptionIndex === undefined
-    ? undefined
-    : `${optionIdPrefix}-${focusedOptionIndex}`;
 
+  const [focusedOptionId, setFocusedOptionId] = useState<string | undefined>(undefined);
   const [latestUserInput, setLatestUserInput] = useState(inputValue);
   const [screenReaderKey, setScreenReaderKey] = useState(0);
 
   const inputRef = useRef<HTMLInputElement>(document.createElement('input'));
+  const dropdownRef = useRef<HTMLDivElement>(document.createElement('div'));
 
   if (!shouldDisplayDropdown && screenReaderKey) {
     setScreenReaderKey(0);
   }
 
+  const childrenArray = React.Children.toArray(children);
+  const childrenWithProps = childrenArray.map((child, index) => {
+    if (!(React.isValidElement(child) && child.type === DropdownSection)) {
+      return child;
+    }
+
+    const modifiedOnClickOption = (option: Option, optionIndex: number) => {
+      child.props.onClickOption?.(option, optionIndex);
+      dispatch({ type: 'HideSections' });
+    }
+
+    const modifiedOnFocusChange = (value: string, focusedOptionId: string) => {
+      child.props.onFocusChange?.(value, focusedOptionId);
+      setFocusedOptionId(focusedOptionId);
+    };
+
+    if (focusedSectionIndex === undefined) {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'reset', onClickOption: modifiedOnClickOption });
+    } else if (index === focusedSectionIndex) {
+      return React.cloneElement(child, {
+        onLeaveSectionFocus, focusStatus: 'active', onFocusChange: modifiedOnFocusChange, onClickOption: modifiedOnClickOption
+      });
+    } else {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'inactive', onClickOption: modifiedOnClickOption });
+    }
+  });
+
+  const numSections = childrenWithProps.length;
+
+  function onLeaveSectionFocus(focusNext: boolean) {
+    if (focusedSectionIndex === undefined && focusNext) {
+      dispatch({ type: 'FocusSection', newIndex: 0 });
+    } else if (focusedSectionIndex !== undefined) {
+      let newSectionIndex: number | undefined = focusNext
+        ? focusedSectionIndex + 1
+        : focusedSectionIndex - 1;
+      if (newSectionIndex < 0) {
+        newSectionIndex = undefined;
+        onInputChange(latestUserInput);
+      } else if (newSectionIndex > numSections - 1) {
+        newSectionIndex = numSections - 1;
+      }
+      dispatch({ type: 'FocusSection', newIndex: newSectionIndex });
+    }
+  }
+
   function handleDocumentClick(evt: MouseEvent) {
     const target = evt.target as HTMLElement;
-    if (!target.isSameNode(inputRef.current)) {
-      dispatch({ type: 'HideOptions' });
+    if (!(target.isSameNode(inputRef.current) || (dropdownRef.current?.contains(target)))) {
+      dispatch({ type: 'HideSections' });
+    }
+  }
+
+  function handleDocumentKeydown(evt: globalThis.KeyboardEvent) {
```

Comment:
I think having both `handleDocumentKeydown` and `onKeyDown` might be a little confusing. Do you think it makes sense to rename `onKeyDown` to `handleInputElementKeydown`?

Reply:
Yes, that would be more clear about which keydown is being handled

Reply:
Updated!